### DISCUSSION TITLE
Vercel: GET 404 Error with Formik library

### DISCUSSION THREAD

**Discussion by Justus-Becker-Rabbenstein** [neuefische/questions#402] [Web] [answered]
Hi,
enviroment: Next.js with Formik library

I installed Formik with npm and implemented it into my code to handle the changes of textareas in pages/myProfile/index.js
On my local installation (npm run dev - localhost:3000) everything works fine.
The deploy on Vercel works fine, too.
However: if you login as 'Max Mustermann' and click on 'Myself' in the burger menu the values are not displayed correctly in the Vercel deployment. In the console you get an GET 404 error hint, that says neccessary values not found.
The Formik form uses a const {handleSubmit, handleChange, values} = useFormik({ and the textareas are supposed to be filled with values with the values from Formik. This appears to be not working on Vercel.
I tryed switching the paths of the _db array around for the images, but that didn't help. I tryed fixing the warning of the useEffect, but it didn't solve anything. I don't know what to do. It seems like the values defined in the useFormik aren't read properly on Vercel, I think.

Vercel Deployment
![vercel](https://user-images.githubusercontent.com/43116175/207838255-29214c31-21b3-4860-a663-b7e2869f6960.jpg)
Local dev server
![local](https://user-images.githubusercontent.com/43116175/207838250-14c97152-7c76-4d66-ba26-9d66fd9415d7.jpg)
Git Pull Request
https://github.com/Justus-Becker-Rabbenstein/contacto/pull/44

**Comment by ThomasKirst** [comment]
Your data doesn´t seem to be ready at the time when you want to paint the user´s information.

The 404 results of an image request:
https://contacto-git-feature-editownprofile-justus-becker-rabbenstein.vercel.app/Awaiting%20data%20...

In your MyProfile page you are passing a string `Awaiting data` into your image component:
```js
  <ContainerProfileImage
    height="50vh"
    width="50vw"
    src={values ? values.image : "Awaiting data ..."}
    alt={values ? values.name : "Awaiting data ..."}
    onClick={handleImgClicked}
  />
```

Apparently, your values are not ready by that time.

It might help to just add a check before rendering anything:
```js
if (!values) return null;
```

This will prevent the rendering of the component when it has no data yet.  When the data becomes available, you should see the image rendered (if your data supports it).

I hope this helps.
– Thomas

**Reply by Justus-Becker-Rabbenstein** [reply]
Hi Thomas, 
I implemened the solution:
`if (!values) {
    return null;
  } else {
    return (
      <>
        <h6>Login/My Profile</h6>
        <ContainerForm onSubmit={handleSubmit}>
          <ContainerLottie>
            <Lottie options={defaultOptions} height="5rem" width="5rem" />
          </ContainerLottie>
          [...]`
This does not fix the problem.
[New Deployment on Vercel](https://contacto-eqcda8bdh-justus-becker-rabbenstein.vercel.app/)
Best regards
Justus

Edit: the solution without the else branch doesn't work either.
[Vercel without else branch](https://contacto-27y7p4syp-justus-becker-rabbenstein.vercel.app/)
`if (!values) {
    return null;
  }
  return (
    <>
      <h6>Login/My Profile</h6>
[...]`

**Comment by ThomasKirst** [comment]
It´s hard to follow along your code on the myProfile page.
It seems like you are handling a login there, `values` seems to be something from formik (do you really need a form library?).
Then you have a `useEffect` which itself listens on `values`, which performs some login logic. Could all of that be extracted to another place (maybe a useLogin custom hook)?
From what I understand, you want to show the user´s information on that page.
If you need `values` and this is somehow managed by formik, which I don´t want to go into, you could also consider conditional rendering:

```js
{ values?.image &&
  <ContainerProfileImage
    height="50vh"
    width="50vw"
    src={values.image}
    alt={values?.name}
    onClick={handleImgClicked}
  />
}
```

**Comment by Justus-Becker-Rabbenstein** [comment, marked as answer]
Hi,
the error on Vercel occured because I used userArray.splice(indexOfUser, 1, values); on a useState array without the setter function. Strangely enough this was okay on my local machine.
I rewrote the code with coach Ralf Siewert to use the setter function in state uplifting instead.
_app.js state uplifting for userArray
`  // State uplifting: setter userarry rewrite user on update
  const handleUpdateUser = updatedUser => {
    setUserArray(
      userArray.map(user => {
        if (user.id === updatedUser.id) {
          return updatedUser;
        } else {
          return user;
        }
      })
    );
  };`
myProfile Formik
`  // this updates the data in the user data array
  // Form Submit Logic with Formik library for React (Textareas)
  const {handleSubmit, handleChange, values} = useFormik({
    initialValues: ownProfileUser,
    enableReinitialize: true,
    onSubmit: values => {
      onSubmitLogin(values.name);
      onUpdateUser(values);
      alert("Own profile successfully updated.");
    },
  });`
